ogame/__init__.py

```diff
@@ -1154,64 +1154,49 @@ def spyreports(self, firstpage=1, lastpage=30):
-                tech_amount = int(li.find('span', 'fright').text.replace('.', ''))
-                if tech_id in [212, 217]:
-                    tech_name = const.buildings.building_name(
-                        (tech_id, None, None)
-                    )
-                else:
-                    tech_name = const.ships.ship_name(
-                        (tech_id, None, 'shipyard')
-                    )
-                spied_fleet.update({tech_name: tech_amount})
-            spied_defences = {}
-            ul = bs4.find('ul', {'data-type': 'defense'})
-            for li in ul.find_all('li'):
-                tech_id = int(re.search(r'([0-9]+)', li.find('img')['class'][0]).group(1))
-                tech_name = const.buildings.defense_name(
-                    (tech_id, None, 'defenses')
-                )
-                tech_amount = int(li.find('span', 'fright').text.replace('.', ''))
-                spied_defences.update({tech_name: tech_amount})
-            spied_buildings = {}
-            ul = bs4.find('ul', {'data-type': 'buildings'})
-            for li in ul.find_all('li'):
-                tech_id = int(re.search(r'([0-9]+)', li.find('img')['class'][0]).group(1))
-                tech_name = const.buildings.building_name(
-                    (tech_id, None, None)
-                )
-                tech_level = int(li.find('span', 'fright').text.replace('.', ''))
-                spied_buildings.update({tech_name: tech_level})
-            spied_research = {}
-            ul = bs4.find('ul', {'data-type': 'research'})
-            for li in ul.find_all('li'):
-                tech_id = int(re.search(r'([0-9]+)', li.find('img')['class'][0]).group(1))
-                tech_name = const.research.research_name(
-                    (tech_id, None, 'research')
-                )
-                tech_amount = int(li.find('span', 'fright').text.replace('.', ''))
-                spied_research.update({tech_name: tech_amount})
+                resources_data[resource_name[0]] = int(resource['title'].replace('.', ''))
+
+            def get_tech_qt(tech_type):
+                tech_list = bs4.find('ul', {'data-type': tech_type})
+                for tech in tech_list.find_all('li'):
+                    tech_id = int(re.search(r'([0-9]+)', tech.find('img')['class'][0]).group(1))
+                    tech_amount = int(tech.find('span', 'fright').text.replace('.', ''))
+                    yield (tech_id, tech_amount)
+
+            spied_data = {'ships': {}, 'defense': {}, 'buildings': {}, 'research': {}}
+            const_data = {
+                'ships': [const.ships.ship_name, 'shipyard'],
+                'defense': [const.buildings.defense_name, 'defenses'],
+                'buildings': [const.buildings.building_name, None],
+                'research': [const.research.research_name, 'research']
+            }
+            for tech_type in spied_data.keys():
```

Comment:
O(n2) complexity.
can this be reduced to O(n)?

Reply:
Sorry, I never understood that O(x) thing...

I looked on Wikipedia, now I sort of understand what O(x) mean but I can't find where O(n<sup>2</sup>) is in my code. Can you please pointing it out so I can better understand?

Reply:
i am pointing out that you have a quite complex algo.
for
    for
responds to a O(n2)
and when possible its best to avoid doing it like this because its quite scales exponential. depends on the runtime it will run against you.

if i have 20 spy messages it will run 20^2=400 times

i try to aboid loops inside loops